rebac-admin-backend/v1/groups.go

```diff
@@ -1,11 +1,241 @@
-type GroupsServiceBackend interface {
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/canonical/identity-platform-admin-ui/rebac-admin-backend/v1/resources"
+)
+
+// GetGroups returns the list of known groups.
+// (GET /groups)
+func (h handler) GetGroups(w http.ResponseWriter, req *http.Request, params resources.GetGroupsParams) {
+	ctx := req.Context()
+
+	groups, err := h.Groups.ListGroups(ctx, &params)
+	if err != nil {
+		response := h.GroupsErrorMapper.MapError(err)
+		writeResponse(w, response.Status, response)
+		return
+	}
+
+	response := resources.GetGroupsResponse{
+		Data:   groups.Data,
+		Status: http.StatusOK,
+	}
+
+	writeResponse(w, http.StatusOK, response)
+}
+
+// PostGroups adds a new group.
+// (POST /groups)
+func (h handler) PostGroups(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	group := new(resources.Group)
+	defer req.Body.Close()
+
+	if err := json.NewDecoder(req.Body).Decode(group); err != nil {
+		writeErrorResponse(w, NewValidationError("request doesn't match the expected schema"))
+		return
+	}
+
+	group, err := h.Groups.CreateGroup(ctx, group)
+	if err != nil {
+		response := h.GroupsErrorMapper.MapError(err)
+		writeResponse(w, response.Status, response)
+		return
+	}
+
+	writeResponse(w, http.StatusCreated, group)
+}
+
+// DeleteGroupsItem deletes the specified group.
+// (DELETE /groups/{id})
+func (h handler) DeleteGroupsItem(w http.ResponseWriter, req *http.Request, id string) {
+	ctx := req.Context()
+
+	_, err := h.Groups.DeleteGroup(ctx, id)
+	if err != nil {
+		response := h.GroupsErrorMapper.MapError(err)
+		writeResponse(w, response.Status, response)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
 }
 
-// GroupsAuthorizationBackend defines an abstract backend to handle authorization for Groups.
-type GroupsAuthorizationBackend interface {
+// GetGroupsItem returns the group identified by the provided ID.
+// (GET /groups/{id})
+func (h handler) GetGroupsItem(w http.ResponseWriter, req *http.Request, id string) {
+	ctx := req.Context()
+
+	identity, err := h.Groups.GetGroup(ctx, id)
+	if err != nil {
+		response := h.GroupsErrorMapper.MapError(err)
+		writeResponse(w, response.Status, response)
+		return
+	}
+
+	writeResponse(w, http.StatusOK, identity)
+}
+
+// PutGroupsItem updates the group identified by the provided ID.
+// (PUT /groups/{id})
+func (h handler) PutGroupsItem(w http.ResponseWriter, req *http.Request, id string) {
+	ctx := req.Context()
+
+	group := new(resources.Group)
+	defer req.Body.Close()
+
+	if err := json.NewDecoder(req.Body).Decode(group); err != nil {
+		writeErrorResponse(w, NewValidationError("request doesn't match the expected schema"))
+		return
+	}
+
+	if id != *group.Id {
+		writeErrorResponse(w, NewValidationError("group ID from path does not match the Group object"))
+		return
+	}
+
+	group, err := h.Groups.UpdateGroup(ctx, group)
+	if err != nil {
+		response := h.GroupsErrorMapper.MapError(err)
```

Comment:
looking at the usage pattern, it seems better to have something like
```
h.GroupsErrorMapper.WriteErrorResponse(err)
return
```

Reply:
I'm a little against this, reason being: I would let the mapper do the mapping work and that work only, following the single responsibility principle. What do you think?

Reply:
Question; are you suggesting this to make it more readable/terse? or it's more about the semantics?

I'm asking this, because the `GroupsErrorMapper` is something that users (i.e., product services) should provide us with. So, changing `MapError` (which is expected to return a `Response` struct instance) to `WriteErrorResponse` (which implies user access to the undelrying `http.ResponseWriter`) affects the way we tend to abstract things. I, personally, rather not share the `http.ResponseWriter` instance with the user.

Anyway, if your suggestion is more about making it terse, I totally agree. I think we can add an internal helper method for this.

Reply:
I think a generic helper method to do these two lines would be a good thing. I'll add it to my other PR (#205).